src/Parser/Core/Modules/SpellUsable.js

```diff
@@ -0,0 +1,247 @@
+import SPELLS from 'common/SPELLS';
+import { formatDuration } from 'common/format';
+
+import Module from 'Parser/Core/Module';
+
+import CastEfficiency from './CastEfficiency';
+
+const debug = false;
+
+function spellName(spellId) {
+  return SPELLS[spellId] ? SPELLS[spellId].name : '???';
+}
+
+class SpellUsable extends Module {
+  static dependencies = {
+    castEfficiency: CastEfficiency,
+  };
+  _currentCooldowns = {};
+
+  /**
+   * Whether the spell can be cast. This is not the opposite of `isOnCooldown`! A spell with 2 charges, 1 available and 1 on cooldown would be both available and on cooldown at the same time.
+   */
+  isAvailable(spellId) {
+    if (this.isOnCooldown(spellId)) {
+      const maxCharges = this.castEfficiency.getMaxCharges(spellId);
+      if (maxCharges > this._currentCooldowns[spellId].chargesOnCooldown) {
+        return true;
+      }
+      return false;
+    } else {
+      return true;
+    }
+  }
+  /**
+   * Whether the spell is on cooldown. If a spell has multiple charges with 1 charge on cooldown and 1 available, this will return `true`. Use `isAvailable` if you just want to know if a spell is castable. Use this if you want to know if a spell is current cooling down, regardless of being able to cast it.
+   */
+  isOnCooldown(spellId) {
+    return !!this._currentCooldowns[spellId];
+  }
+  /**
+   * Returns the amount of time remaining on the cooldown.
+   * @param {number} spellId
+   * @param {number} timestamp Override the timestamp if it may be different from the current timestamp.
+   * @returns {number|null} Returns null if the spell isn't on cooldown.
+   */
+  cooldownRemaining(spellId, timestamp = this.owner.currentTimestamp) {
+    if (!this.isOnCooldown(spellId)) {
+      return null;
```

Comment:
What is the reason this is null rather than 0? Return type consistency would be nice here.

Reply:
I want consumers of the function to take it being off cooldown into account when they're calling this method so they're less likely to create mistakes. This should probably throw instead.